contributors/devel/sig-instrumentation/migration-to-structured-logging.md

```diff
@@ -0,0 +1,346 @@
+func KRef(namespace, name string) ObjectRef
+
+// ObjectRef represents a reference to a kubernetes object used for logging purpose
+// In text logs it is serialized into "{namespace}/{name}" or "{name}" if namespace is empty
+type ObjectRef struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// KMetadata is a subset of the kubernetes k8s.io/apimachinery/pkg/apis/meta/v1.Object interface
+// this interface may expand in the future, but will always be a subset of the
+// kubernetes k8s.io/apimachinery/pkg/apis/meta/v1.Object interface
+type KMetadata interface {
+	GetName() string
+	GetNamespace() string
+}
+```
+
+## Migration
+
+1. Change log functions to structured equivalent
+1. Remove string formatting from log message
+1. Name arguments
+1. Use `klog.KObj` and `klog.KRef` for Kubernetes object references
+1. Verify log output
+
+## Change log functions to structured equivalent
+
+Structured logging functions follow a different logging interface design than other functions in `klog`. They follow
+minimal design from [logr] thus there is no one-to-one mapping.
+
+Simplified mapping between functions:
+* `klog.Infof`, `klog.Info`, `klog.Infoln`, `klog.InfoDepth` -> `klog.InfoS`
+* `klog.V(N).Infof`, `klog.V(N).Info`, `klog.V(N).Infoln` -> `klog.V(N).InfoS`
+* `klog.Warning`, `klog.Warningf`, `klog.Warningln`, `klog.WarningDepth` -> `klog.InfoS`
+* `klog.Error`, `klog.Errorf`, `klog.Errorln`, `klog.ErrorDepth` -> `klog.ErrorS`
+* `klog.Fatal`, `klog.Fatalf`, `klog.Fatalln`, `klog.FatalDepth` -> `klog.ErrorS`
+
+### Removing Depth
+
+Functions with depth (`klog.InfoDepth`, `klog.WarningDepth`, `klog.ErrorDepth`, `klog.FatalDepth`) are used to indicate
+that the source of the log (added as metadata in log) is different than the invocation of logging library. This is
+usually used when implementing logging util functions. In those situations util functions should be removed and `klog`
+call can be moved up the call stack.
+
+For example
+```go
+func Handle(w http.ReponseWriter, r *http.Request) {
+    logHTTPRequest(r)
+    handle(w, r)
+}
+func logHTTPRequest(r *http.Request) {
+    klog.InfoDepth(1, "Received HTTP request")
+}
+```
+should be replaced with
+```go
+func Handle(w http.ReponseWriter, r *http.Request) {
+    klog.Info("Received HTTP request")
+    handle(w, r)
+}
+```
+
+### Using ErrorS
+
+With `klog` structured logging borrowing the interface from [logr] it also inherits it's differences in semantic of
+error function. This is visible in `klog.ErrorS` which requires passing an `error` as the first argument. This is
+because in [logr] Error is used to inform users about errors received from subordinate function calls.
+This means that `klog.ErrorS` should only be used when there is a obvious Golang `error` object available.
+
+Creating an error just to call `klog.ErrorS` or passing `nil` is discouraged, please consider using `klog.InfoS` instead.
```

Comment:
Makes sense, I will apply your suggestion.

Reply:
It seems that the current implementation of klog.ErrorS() calls logr.Info() if the err is nil. If we want to encourage people to pass nil I believe that this should be changed so that logr.Error() would be called.

Reply:
Added https://github.com/kubernetes/klog/issues/153

Reply:
One note that `klog.ErrorS` is not guaranteed to generate a stacktrace as by default it still calls `klog.print`

Reply:
Yep, we just kinda note in logr that `Error` may take special action to deal with errors in logr (see below).  For instance, I could see an impl that formatted errors for capture by special error recording tools (e.g. stackdriver error reporting or somesuch), ones that print stack traces (like zap), and so on.  It's nice because it leaves us a way to do cool stuff like that in the future, even if we don't do it by default now.

---

FWIW, logr docs say:

> Furthermore, certain implementations may choose to attach additional information (such as stack traces) on calls to Error, so it's preferred to use Error to log errors.